packages/back-end/src/models/EventModel.ts

```diff
@@ -90,62 +106,174 @@ const eventSchema = new mongoose.Schema({
-      data: data,
+      data: { ...payload, api_version: API_VERSION, created: Date.now() },
     });
 
-    return toInterface(doc) as EventInterface<NotificationEvent>;
+    const event = toInterface(doc) as BaseEventInterface<
+      NotificationEventPayload<Resource, Event>,
+      typeof MODEL_VERSION
+    >;
+
+    new EventNotifier(event.id).perform();
   } catch (e) {
     logger.error(e);
-    return null;
   }
 };
 
+// createEvent can handle creating the diff
+
+export type CreateEventData<
+  Resource extends NotificationEventResource,
+  Event extends ResourceEvents<Resource>,
+  Payload = NotificationEventPayloadSchemaType<Resource, Event>
+> = NotificationEventPayloadDataType<Resource, Event, Payload, Payload>;
+
+export const hasPreviousAttributes = <
+  Resource extends NotificationEventResource,
+  Event extends ResourceEvents<Resource>,
+  Payload = NotificationEventPayloadSchemaType<Resource, Event>
+>(
+  data: CreateEventData<Resource, Event, Payload>
+): data is {
+  object: Payload;
+  previous_attributes: Payload;
+} & NotificationEventPayloadExtraAttributes<Resource, Event> =>
+  Object.keys(data).includes("previous_attributes");
+
+const diffData = <
+  Resource extends NotificationEventResource,
+  Event extends ResourceEvents<Resource>,
+  Payload = NotificationEventPayloadSchemaType<Resource, Event>
+>(
+  data: CreateEventData<Resource, Event, Payload>
+): NotificationEventPayloadDataType<Resource, Event, Payload> => {
+  if (!hasPreviousAttributes(data)) return data;
+
+  const { object, previous_attributes } = data as {
+    object: Record<string, unknown>;
+    previous_attributes: Record<string, unknown>;
+  };
+
+  return ({
+    object,
+    previous_attributes: [
+      ...new Set([
+        ...Object.keys(object),
+        ...Object.keys(previous_attributes as object),
+      ]),
+    ].reduce(
+      (diff, key) => ({
+        ...diff,
+        ...(isEqual(object[key], previous_attributes[key])
+          ? {}
+          : { [key]: previous_attributes[key] }),
+      }),
+      {}
+    ),
+  } as unknown) as NotificationEventPayloadDataType<Resource, Event, Payload>;
+};
+
+export type CreateEventParams<
+  Resource extends NotificationEventResource,
+  Event extends ResourceEvents<Resource>,
+  Payload = NotificationEventPayloadSchemaType<Resource, Event>
+> = {
+  context: ReqContext;
+  object: Resource;
+  event: Event;
+  data: CreateEventData<Resource, Event, Payload>;
+  containsSecrets: boolean;
+  projects: string[];
+  tags: string[];
+  environments: string[];
+};
+
+export const createEvent = async <
+  Resource extends NotificationEventResource,
+  Event extends ResourceEvents<Resource>
+>({
+  context,
+  object,
+  event,
+  data,
+  containsSecrets,
+  projects,
+  tags,
+  environments,
+}: CreateEventParams<Resource, Event>) =>
+  createEventWithPayload<Resource, Event>({
+    payload: {
+      event: `${object}.${event}`,
+      object,
+      data: diffData(data),
```

Comment:
Right now, we expect people to pass in the full previous object as `previous_attributes` and then we rewrite it here to be a filtered version with only the changed keys.  I don't know if this is super intuitive and might be too much magic happening internally.

I think it might be clearer if we're a little more explicit, something like this:

```ts
createEvent({
   data: {
      object: feature,
      previous_attributes: getPreviousAttributesFromDiff(feature, prevFeature)
   },
   ...
})
```

Basically, just bring this diff logic up to where `createEvent` is being called instead of hiding it internally.

Reply:
I usually prefer encapsulating everything to prevent crazy node dependencies and also make refactorization/reduce boilerplate. Happy to do that, tho, I understand the other side of this argument.